clang c++ support

Thanks to @MrAnno 's recent PR (https://github.com/syslog-ng/syslog-ng/pull/4924), C++ / GRPC modules compile again on FreeBSD. However, using GCC's libstdc++ is still hard-coded everywhere, so linking fails.

Currently I use the workaround:
```
for i in `find . -type f` ; do gsed -i 's/stdc++/c++/g' $i ; done
```
Without it:

```
libtool: link: c++  -fPIC -DPIC -shared -nostdlib /usr/lib/crti.o /usr/lib/crtbeginS.o  modules/grpc/protos/.libs/libgrpc_protos_la-apphook.o ./modules/grpc/protos/otlp/opentelemetry/proto/collector/logs/v1/.libs/modules_grpc_protos_libgrpc_protos_la-logs_service.grpc.pb.o ./modules/grpc/protos/otlp/opentelemetry/proto/collector/logs/v1/.libs/modules_grpc_protos_libgrpc_protos_la-logs_service.pb.o ./modules/grpc/protos/otlp/opentelemetry/proto/collector/metrics/v1/.libs/modules_grpc_protos_libgrpc_protos_la-metrics_service.grpc.pb.o ./modules/grpc/protos/otlp/opentelemetry/proto/collector/metrics/v1/.libs/modules_grpc_protos_libgrpc_protos_la-metrics_service.pb.o ./modules/grpc/protos/otlp/opentelemetry/proto/collector/trace/v1/.libs/modules_grpc_protos_libgrpc_protos_la-trace_service.grpc.pb.o ./modules/grpc/protos/otlp/opentelemetry/proto/collector/trace/v1/.libs/modules_grpc_protos_libgrpc_protos_la-trace_service.pb.o ./modules/grpc/protos/otlp/opentelemetry/proto/common/v1/.libs/modules_grpc_protos_libgrpc_protos_la-common.pb.o ./modules/grpc/protos/otlp/opentelemetry/proto/logs/v1/.libs/modules_grpc_protos_libgrpc_protos_la-logs.pb.o ./modules/grpc/protos/otlp/opentelemetry/proto/metrics/v1/.libs/modules_grpc_protos_libgrpc_protos_la-metrics.pb.o ./modules/grpc/protos/otlp/opentelemetry/proto/resource/v1/.libs/modules_grpc_protos_libgrpc_protos_la-resource.pb.o ./modules/grpc/protos/otlp/opentelemetry/proto/trace/v1/.libs/modules_grpc_protos_libgrpc_protos_la-trace.pb.o ./modules/grpc/protos/loki/.libs/modules_grpc_protos_libgrpc_protos_la-push.grpc.pb.o ./modules/grpc/protos/loki/.libs/modules_grpc_protos_libgrpc_protos_la-push.pb.o ./modules/grpc/protos/google-proto/google/api/.libs/modules_grpc_protos_libgrpc_protos_la-annotations.pb.o ./modules/grpc/protos/google-proto/google/api/.libs/modules_grpc_protos_libgrpc_protos_la-client.pb.o ./modules/grpc/protos/google-proto/google/api/.libs/modules_grpc_protos_libgrpc_protos_la-field_behavior.pb.o ./modules/grpc/protos/google-proto/google/api/.libs/modules_grpc_protos_libgrpc_protos_la-http.pb.o ./modules/grpc/protos/google-proto/google/api/.libs/modules_grpc_protos_libgrpc_protos_la-launch_stage.pb.o ./modules/grpc/protos/google-proto/google/api/.libs/modules_grpc_protos_libgrpc_protos_la-resource.pb.o ./modules/grpc/protos/google-proto/google/cloud/bigquery/storage/v1/.libs/modules_grpc_protos_libgrpc_protos_la-annotations.pb.o ./modules/grpc/protos/google-proto/google/cloud/bigquery/storage/v1/.libs/modules_grpc_protos_libgrpc_protos_la-arrow.pb.o ./modules/grpc/protos/google-proto/google/cloud/bigquery/storage/v1/.libs/modules_grpc_protos_libgrpc_protos_la-avro.pb.o ./modules/grpc/protos/google-proto/google/cloud/bigquery/storage/v1/.libs/modules_grpc_protos_libgrpc_protos_la-protobuf.pb.o ./modules/grpc/protos/google-proto/google/cloud/bigquery/storage/v1/.libs/modules_grpc_protos_libgrpc_protos_la-storage.grpc.pb.o ./modules/grpc/protos/google-proto/google/cloud/bigquery/storage/v1/.libs/modules_grpc_protos_libgrpc_protos_la-storage.pb.o ./modules/grpc/protos/google-proto/google/cloud/bigquery/storage/v1/.libs/modules_grpc_protos_libgrpc_protos_la-stream.pb.o ./modules/grpc/protos/google-proto/google/cloud/bigquery/storage/v1/.libs/modules_grpc_protos_libgrpc_protos_la-table.pb.o ./modules/grpc/protos/google-proto/google/rpc/.libs/modules_grpc_protos_libgrpc_protos_la-status.pb.o   -Wl,-rpath -Wl,/root/freebsd/syslog-ng4-devel/work/syslog-ng-4.7.1.128.g88ecd1e/lib/.libs -Wl,-rpath -Wl,/usr/local/lib -L/usr/local/lib ./lib/.libs/libsyslog-ng.so -lprotobuf -labsl_log_internal_check_op -labsl_leak_check -labsl_die_if_null -labsl_log_internal_conditions -labsl_log_internal_message -labsl_examine_stack -labsl_log_internal_format -labsl_log_internal_proto -labsl_log_internal_nullguard -labsl_log_internal_log_sink_set -labsl_log_sink -labsl_log_entry -labsl_log_initialize -labsl_log_globals -labsl_log_internal_globals -lutf8_validity -lutf8_range -lpthread -lgrpc++ -lgrpc -labsl_statusor -lgpr -labsl_flags -labsl_flags_internal -labsl_flags_reflection -labsl_flags_private_handle_accessor -labsl_flags_commandlineflag -labsl_flags_commandlineflag_internal -labsl_flags_config -labsl_flags_program_name -labsl_hash -labsl_city -labsl_low_level_hash -labsl_raw_hash_set -labsl_hashtablez_sampler -labsl_flags_marshalling -labsl_random_distributions -labsl_random_seed_sequences -labsl_random_internal_pool_urbg -labsl_random_internal_randen -labsl_random_internal_randen_hwaes -labsl_random_internal_randen_hwaes_impl -labsl_random_internal_randen_slow -labsl_random_internal_platform -labsl_random_internal_seed_material -labsl_random_seed_gen_exception -labsl_status -labsl_cord -labsl_cordz_info -labsl_cord_internal -labsl_cordz_functions -labsl_exponential_biased -labsl_cordz_handle -labsl_crc_cord_state -labsl_crc32c -labsl_crc_internal -labsl_crc_cpu_detect -labsl_bad_optional_access -labsl_strerror -labsl_str_format_internal -labsl_synchronization -labsl_graphcycles_internal -labsl_stacktrace -labsl_symbolize -labsl_debugging_internal -labsl_demangle_internal -labsl_malloc_internal -labsl_time -labsl_civil_time -labsl_strings -labsl_strings_internal -lrt -labsl_base -labsl_spinlock_wait -labsl_int128 -labsl_throw_delegate -labsl_time_zone -labsl_bad_variant_access -labsl_raw_logging_internal -labsl_log_severity -lstdc++ -L/usr/lib -lc++ -lm -lc -lgcc -lgcc_s /usr/lib/crtendS.o /usr/lib/crtn.o  -O2 -g -O2 -fstack-protector-strong -fstack-protector-strong   -pthread -Wl,-soname -Wl,libgrpc-protos.so.0 -o modules/grpc/protos/.libs/libgrpc-protos.so.0.0.0
c++: warning: argument unused during compilation: '-pthread' [-Wunused-command-line-argument]
ld: error: unable to find library -lstdc++
c++: error: linker command failed with exit code 1 (use -v to see invocation)
gmake[2]: *** [Makefile:16021: modules/grpc/protos/libgrpc-protos.la] Error 1
gmake[1]: *** [Makefile:31019: all-recursive] Error 1
gmake: *** [Makefile:13288: all] Error 2
*** Error code 1
```

--- a/Makefile.am
+++ b/Makefile.am
@@ -126,7 +126,7 @@ endif
 
 TEST_CFLAGS		= -DTOP_SRCDIR=\"$(abs_top_srcdir)\" $(AM_CFLAGS) $(CRITERION_CFLAGS)
 
-TEST_CXXFLAGS		= -DTOP_SRCDIR=\"$(abs_top_srcdir)\" $(AM_CXXFLAGS) $(CRITERION_CFLAGS) -lstdc++
+TEST_CXXFLAGS		= -DTOP_SRCDIR=\"$(abs_top_srcdir)\" $(AM_CXXFLAGS) $(CRITERION_CFLAGS)
 
 TEST_LDADD		= $(LIBTEST_LIBS) $(CRITERION_LIBS) \
 			  $(top_builddir)/lib/libsyslog-ng.la \

--- a/lib/compat/cpp-start.h
+++ b/lib/compat/cpp-start.h
@@ -49,9 +49,10 @@
  *     and plugin.c.  This minimizes the chance of including an incompatible
  *     C header.
  *   - Build/link the C++ code as C++ separately and link to that from your
- *     C lib.  Adding -lstdc++ to LIBADD is neccessary in this case as it is
- *     automatically added while linking the C++ object itself, but not when
- *     linking to the C++ object from a C lib.
+ *     C lib. In case your C++ library is static, add the following to the
+ *     the C lib to force linking against the appropriate C++ standard
+ *     library (libc++, libstdc++):
+ *       nodist_EXTRA_*_SOURCES = force-cpp-linker-with-default-stdlib.cpp
  *   - In your C++ code it is not possible to derive from a C class in the
  *     usual C way by adding a super field and filling its free_fn, because
  *     we do our own reference counting and freeing logic and we cannot rely

--- a/modules/cloud-auth/Makefile.am
+++ b/modules/cloud-auth/Makefile.am
@@ -40,6 +40,8 @@ modules_cloud_auth_libcloud_auth_la_LIBADD = \
   $(MODULE_DEPS_LIBS) \
   $(top_builddir)/modules/cloud-auth/libcloud_auth_cpp.la
 
+nodist_EXTRA_modules_cloud_auth_libcloud_auth_la_SOURCES = force-cpp-linker-with-default-stdlib.cpp
+
 modules_cloud_auth_libcloud_auth_la_LDFLAGS = \
   $(MODULE_LDFLAGS)
 

--- a/modules/examples/Makefile.am
+++ b/modules/examples/Makefile.am
@@ -35,7 +35,7 @@ EXTRA_modules_examples_libexamples_la_DEPENDENCIES = $(MODULE_DEPS_LIBS) $(EXAMP
 modules_examples_libexamples_la_LDFLAGS = $(MODULE_LDFLAGS)
 
 if ENABLE_CPP
-modules_examples_libexamples_la_LIBADD += -lstdc++
+nodist_EXTRA_modules_examples_libexamples_la_SOURCES = force-cpp-linker-with-default-stdlib.cpp
 endif
 
 EXTRA_DIST += modules/examples/CMakeLists.txt

--- a/modules/grpc/bigquery/Makefile.am
+++ b/modules/grpc/bigquery/Makefile.am
@@ -41,8 +41,9 @@ modules_grpc_bigquery_libbigquery_la_CPPFLAGS = \
 modules_grpc_bigquery_libbigquery_la_LIBADD = \
   $(MODULE_DEPS_LIBS) \
   $(top_builddir)/modules/grpc/protos/libgrpc-protos.la \
-  $(top_builddir)/modules/grpc/bigquery/libbigquery_cpp.la \
-  -lstdc++
+  $(top_builddir)/modules/grpc/bigquery/libbigquery_cpp.la
+
+nodist_EXTRA_modules_grpc_bigquery_libbigquery_la_SOURCES = force-cpp-linker-with-default-stdlib.cpp
 
 modules_grpc_bigquery_libbigquery_la_LDFLAGS = $(MODULE_LDFLAGS)
 EXTRA_modules_grpc_bigquery_libbigquery_la_DEPENDENCIES = \

--- a/modules/grpc/loki/Makefile.am
+++ b/modules/grpc/loki/Makefile.am
@@ -43,8 +43,9 @@ modules_grpc_loki_libloki_la_CPPFLAGS = \
 modules_grpc_loki_libloki_la_LIBADD = \
   $(MODULE_DEPS_LIBS) \
   $(top_builddir)/modules/grpc/protos/libgrpc-protos.la \
-  $(top_builddir)/modules/grpc/loki/libloki_cpp.la \
-  -lstdc++
+  $(top_builddir)/modules/grpc/loki/libloki_cpp.la
+
+nodist_EXTRA_modules_grpc_loki_libloki_la_SOURCES = force-cpp-linker-with-default-stdlib.cpp
 
 modules_grpc_loki_libloki_la_LDFLAGS = $(MODULE_LDFLAGS)
 EXTRA_modules_grpc_loki_libloki_la_DEPENDENCIES = \

--- a/modules/grpc/otel/Makefile.am
+++ b/modules/grpc/otel/Makefile.am
@@ -61,8 +61,9 @@ modules_grpc_otel_libotel_la_LIBADD = \
   $(MODULE_DEPS_LIBS) \
   $(top_builddir)/modules/grpc/protos/libgrpc-protos.la \
   $(top_builddir)/modules/grpc/otel/libotel_cpp.la \
-  $(top_builddir)/modules/grpc/otel/filterx/libfilterx.la \
-  -lstdc++
+  $(top_builddir)/modules/grpc/otel/filterx/libfilterx.la
+
+nodist_EXTRA_modules_grpc_otel_libotel_la_SOURCES = force-cpp-linker-with-default-stdlib.cpp
 
 modules_grpc_otel_libotel_la_LDFLAGS = $(MODULE_LDFLAGS)
 

--- a/modules/grpc/protos/Makefile.am
+++ b/modules/grpc/protos/Makefile.am
@@ -128,7 +128,8 @@ if ENABLE_EXTRA_WARNINGS
 modules_grpc_protos_libgrpc_protos_la_CXXFLAGS += -Wno-switch-default
 endif
 
-modules_grpc_protos_libgrpc_protos_la_LIBADD = $(MODULE_DEPS_LIBS) $(PROTOBUF_LIBS) $(GRPCPP_LIBS) -lstdc++
+modules_grpc_protos_libgrpc_protos_la_LIBADD = $(MODULE_DEPS_LIBS) $(PROTOBUF_LIBS) $(GRPCPP_LIBS)
+nodist_EXTRA_modules_grpc_protos_libgrpc_protos_la_SOURCES = force-cpp-linker-with-default-stdlib.cpp
 EXTRA_modules_grpc_protos_libgrpc_protos_la_DEPENDENCIES = $(MODULE_DEPS_LIBS)
 
 CLEANFILES += \

--- a/news/other-4933.md
+++ b/news/other-4933.md
@@ -1,0 +1,3 @@
+`bigquery()`, `loki()`, `opentelemetry()`, `cloud-auth()`: C++ modules can be compiled with clang
+
+Compiling and using these C++ modules are now easier on FreeBSD and macOS.

